typing/uniqueness_analysis.ml

```diff
@@ -0,0 +1,1302 @@
+       whole list needs to be forced borrowed. *)
+    | MaybeShared of (unique_barrier ref * Occurrence.t) list
+
+  let to_string = function
+    | Borrowed _ -> "borrowed"
+    | MaybeShared _ -> "maybe_shared"
+
+  let extract_occurrence = function
+    | Borrowed occ -> occ
+    | MaybeShared ((_, occ) :: _) -> occ
+    | MaybeShared [] -> assert false
+
+  let par t0 t1 =
+    match (t0, t1) with
+    | Borrowed _, t | t, Borrowed _ -> t
+    | MaybeShared l0, MaybeShared l1 -> MaybeShared (l0 @ l1)
+
+  let seq t0 t1 =
+    match (t0, t1) with
+    | Borrowed _, t -> t
+    | MaybeShared s, Borrowed _ -> MaybeShared s
+    | MaybeShared l0, MaybeShared l1 -> MaybeShared (l0 @ l1)
+end
+
+module Usage = struct
+  (* We have Unused (top) > Borrowed > Shared > Unique > Error (bot).
+
+     - Unused means unused
+     - Borrowed means read-only access confined to a region
+     - Shared means read-only access that may escape a region. For example,
+     storing the value in a cell that can be accessed later.
+     - Unique means accessing the value as if it's the only pointer. Example
+     includes overwriting.
+     - Error means error happens when composing usage.
+
+     Some observations:
+     - It is sound to relax mode towards Error. It grants the access more
+     "capability" and usually helps performance.
+       For example, relaxing borrowed to shared allows code motion of
+       projections. Relaxing shared to unique allows in-place update.
+
+       An example of the relaxing borrowed to shared:
+
+       let x = r.a in
+       (a lot of codes)
+       x
+
+       In first line, r.memory_address is accessed as borrowed. But if we weaken
+       it to shared and it still mode checks, that means
+       - there is no "unique" access in the "a lot of codes"
+       - or equivalently, that r.memory_address stays unchanged and safe to read
+
+       and as a result, we can delay the projection at `x`.
+
+       The downside of relaxing is the loss of completeness: if we relax too
+       much the program will fail type check. In the extreme case we relax it to
+       Error which fails type check outright (and extremely sound, hehe).
+
+     - The purpose of this uniqueness analysis is to figure out the most relaxed
+     mode for each use, such that we get the best performance, while still
+     type-check. Currently there are really only two choices worth figuring out,
+     Namely
+     - borrowed or shared?
+     - shared or unique?
+
+     As a result, instead of having full-range inference, we only care about the
+     following ranges:
+     - unused
+     - borrowed (Currently not useful, because we don't have explicit borrowing)
+     - borrowed or shared
+     - shared
+     - shared or unique
+     - error
+
+     error is represented as exception which is just easier.
+  *)
+  type t =
+    | Unused
+    (* no good reason to factor out the two, except just to separating the code
+    *)
```

Comment:
I will merge everything - I don't think `MaybeShared` and `MaybeBorrow` are worth their module, because they are closed under `seq`. For example, `seq MaybeShared MaybeShared` = `Shared`.